torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py

```diff
@@ -0,0 +1,818 @@
+# coding=utf-8
+# Copyright 2022 EleutherAI The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch GPTNeoX model."""
+
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.utils.checkpoint
+from torch import nn
+
+from transformers.activations import ACT2FN
+from transformers.file_utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    replace_return_docstrings,
+)
+from transformers.modeling_outputs import BaseModelOutputWithPast, CausalLMOutputWithPast
+from transformers.modeling_utils import PreTrainedModel
+from transformers.utils import logging
+from transformers.models.gpt_neox.configuration_gpt_neox import GPTNeoXConfig
+
+from neuronx_distributed.parallel_layers.layers import ParallelEmbedding, ColumnParallelLinear, RowParallelLinear
+from neuronx_distributed.parallel_layers.loss_functions import parallel_cross_entropy
+from neuronx_distributed.parallel_layers.parallel_state import get_tensor_model_parallel_size
+import neuronx_distributed.parallel_layers.utils as neuronx_dist_utils
+from neuronx_distributed.parallel_layers import move_model_to_device, mappings, layer_norm
+import torch_xla.core.xla_model as xm
+
+from functools import partial
+def _init_normal(std, w):
+    return nn.init.normal_(w, mean=0.0, std=std)
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "trl-internal-testing/tiny-random-GPTNeoXForCausalLM"
+_REAL_CHECKPOINT_FOR_DOC = "EleutherAI/gpt-neox-20b"
+_CONFIG_FOR_DOC = "GPTNeoXConfig"
+
+GPT_NEOX_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "EleutherAI/gpt-neox-20b",
+    # See all GPTNeoX models at https://huggingface.co/models?filter=gpt_neox
+]
+
+
+class GPTNeoXPreTrainedModel(PreTrainedModel):
```

Comment:
Did we change anything from Transformer's NeoX class?

Reply:
We changed a lot:
- replace linears in attention and mlp
- replace embedding and lm_head
- use parallel cross entropy
- modifications for sequence parallel and mask creation on-the-fly

Reply:
Have cleaned up the code by deleting unnecessary codes.